speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/SQLDatabaseStorage.cpp

```diff
@@ -1,229 +1,487 @@
+    EXECUTE_QUERY("SELECT LAST_INSERT_ID()");
 
-    // avoid memory leak
-    sqlite3_free(zErrMsg);
+    int trialId;
+    SELECT_LAST_ID(trialId);
 
-    sqlite3_close(database);
-}
\ No newline at end of file
+    return trialId;
+}
+
+/// @brief Inserts data that is different for each tick in one trial, included tick, user input, gamestate, time step, interventions, and decisions
+/// @param p_trialId id of trial that the simulation is linked with
+void SQLDatabaseStorage::InsertSimulationData(const int p_trialId)
+{
+    while (!m_inputFile.eof()) {
+
+        // tick
+        std::string tick;
+        READ_INPUT(tick);
+
+        // user input
+        std::string steer;
+        std::string brake;
+        std::string gas;
+        std::string clutch;
+
+        READ_INPUT(steer);
+        READ_INPUT(brake);
+        READ_INPUT(gas);
+        READ_INPUT(clutch);
+
+        m_values = "'" + steer + "','" + brake + "','" + gas + "','" + clutch + "'";
+        EXECUTE(INSERT_INTO("userinput", "steer, brake, gas, clutch", m_values));
+
+        int userInputId;
+        SELECT_LAST_ID(userInputId);
+
+        // gamestate
+        std::string x;
+        std::string y;
+        std::string z;
+        std::string directionX;
+        std::string directionY;
+        std::string directionZ;
+        std::string speed;
+        std::string acceleration;
+        std::string gear;
+
+        READ_INPUT(x);
+        READ_INPUT(y);
+        READ_INPUT(z);
+        READ_INPUT(directionX);
+        READ_INPUT(directionY);
+        READ_INPUT(directionZ);
+        READ_INPUT(speed);
+        READ_INPUT(acceleration);
+        READ_INPUT(gear);
+
+        m_values = "'";
+        m_values.append(x);
+        m_values.append("', '");
+        m_values.append(y);
+        m_values.append("', '");
+        m_values.append(z);
+        m_values.append("', '");
+        m_values.append(directionX);
+        m_values.append("', '");
+        m_values.append(directionY);
+        m_values.append("', '");
+        m_values.append(directionZ);
+        m_values.append("', '");
+        m_values.append(speed);
+        m_values.append("', '");
+        m_values.append(acceleration);
+        m_values.append("', '");
+        m_values.append(gear);
+        m_values.append("', '");
+        m_values.append(std::to_string(userInputId));
+        m_values.append("'");
+
+        EXECUTE(INSERT_INTO("gamestate", "x, y, z, direction_x, direction_y, direction_z, speed, acceleration, gear, user_input_id", m_values));
+
+        int gamestateId;
+        SELECT_LAST_ID(gamestateId);
+
+        // time step
+        m_values = "'" + std::to_string(p_trialId) + "','" + tick + "','" + std::to_string(gamestateId) + "'";
+        EXECUTE(INSERT_INTO("timestep", "trial_id, tick, game_state_id", m_values));
+
+        // interventions
+        std::string decision;
+        READ_INPUT(decision);
```

Comment:
You do not account for the possibility of intervention data not being saved (it can be disabled)